### ISSUE TITLE
kubectl documentation does not mention POD_NAMESPACE or inside-kubernetes-cluster mode

### ISSUE THREAD

**Issue by rittneje** [kubernetes/website#26936]
**This is a Bug Report**

**Problem:**

Apparently `kubectl` has two hidden features with respect to what it uses as the default namespace for commands:
1. If `POD_NAMESPACE` environment variable is set, then it uses that.
2. If `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` environment variables are set, and there is a service account token file mounted at "/var/run/secrets/kubernetes.io/serviceaccount/token", then it uses its pod's namespace by default.

These were not clearly documented, resulting in confusion that lead to https://github.com/kubernetes/kubernetes/issues/99884

**Proposed Solution:**

Add this information to the `kubectl` documentation.

**Page to Update:**
https://kubernetes.io/docs/reference/kubectl/overview/ (or one of it siblings)

**Comment by eddiezane** [comment]
This might be coming directly from client-go.

/triage accepted

**Comment by eddiezane** [comment]
Yea this is coming from clientcmd.

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L582

This might help https://github.com/kubernetes/kubectl/issues/927#issuecomment-693483957.

**Comment by geoffcline** [comment]
/assign @geoffcline 

teamwork makes the dreamwork @eddiezane

**Comment by geoffcline** [comment]
> If KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT environment variables are set, and there is a service account token file mounted at "/var/run/secrets/kubernetes.io/serviceaccount/token", then it uses its pod's namespace by default.

can someone explain this just a bit? 

This is what I'm getting

> If you are ssh'd into a pod (as defined by having these env variables defined, and a service acct token), then the cli uses the pod's namespace by default.

**Comment by soltysh** [comment]
> can someone explain this just a bit?

I wrote it down in https://developers.redhat.com/blog/2020/11/20/kubectl-developer-tips-for-the-kubernetes-command-line# 
@geoffcline hope that might clear it a bit :)

**Comment by BenTheElder** [comment]
The original post in this thread also covers what the envs are a bit.

IMHO the two things that trip users up here to be well aware of are:
- 1) If you are using kubectl in a way that you depend on the default namespace (none of your yaml or commands specify the namespace), it will no longer be the default you typically have if that was unset, it will instead be the namespace of the pod you are running in.
- 2) If you meant to interact with another cluster and fail to configure access (e.g. due to some bug in your tooling) you may accidentally try to instead run commands against the cluster you are in

Also probably useful for documenting this: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server

It is possible to disable this behavior by configuring either the service account in the cluster or the pod with `automountServiceAccountToken: false`. See also this thread https://github.com/kubernetes/kubernetes/issues/57601

**Comment by rittneje** [comment]
> It is possible to disable this behavior by configuring either the service account in the cluster or the pod with automountServiceAccountToken: false.

Of course, this doesn't really help if another container (perhaps one injected via a hook) is expecting the token file.
